### DISCUSSION TITLE
PowerShell script with input parameters

### DISCUSSION THREAD

**Discussion by nkrishnakishor** [pester/Pester#2394] [General]
I'm trying to test a PowerShell script which verifies if the required parameters are passed or not before proceeding with rest.

How can I test this using Pester, please help me 🙏

**Comment by fflaten** [comment]
If you're using `[Parameter(Mandatory)]` then you'd likely have to invoke the scripts (or even Pester itself) in a PowerShell/pwsh process launched with `-NonInteractive`. 

That will make it throw an exception for missing parameters instead of prompting for them. AFAIK there's no way to disable the prompt during runtime in a non-disruptive way.

```powershell
# RequiredParameters.ps1
param(
    [Parameter(Mandatory)]
    [string]$Param1,
    [Parameter(Mandatory)]
    [string]$Param2,
    [string]$Param3
)

'Success'

# RequiredParameters.tests.ps1
Describe 'd' {
    BeforeAll {
        $scriptPath = "$PSScriptRoot\RequiredParameters.ps1"
    }
    It 'All required parameters' {
        & $scriptPath -Param1 'a' -Param2 'b' -Param3 'c' | Should -Be 'Success'
    }

    It 'Some required parameters' {
        { & $scriptPath -Param1 'a' -Param3 'c' } | Should -Throw '*missing mandatory parameters: Param2.'
    }

    It 'No required parameters' {
        { & $scriptPath -Param3 'c' } | Should -Throw '*missing mandatory parameters: Param1 Param2.'
    }
}

# Demo
pwsh -NonInteractive -Command "Import-Module Pester; Invoke-Pester '/workspaces/Pester/Samples/RequiredParameters.tests.ps1'"

Starting discovery in 1 files.
Discovery found 3 tests in 179ms.
Running tests.
[+] /workspaces/Pester/Samples/RequiredParameters.tests.ps1 577ms (173ms|254ms)
Tests completed in 590ms
Tests Passed: 3, Failed: 0, Skipped: 0 NotRun: 0
```

Do you really need to test this behavior? If so, you might be better off replacing `[Parameter(Mandatory)]` with e.g. `[ValidateNotNullOrEmpty()]` or your own value check `if ($null -ne $Param1) { throw 'oh no, missing -Param1'}` at the start of the script.

**Reply by nohwnd** [reply]
I was thinking we might replace the host with our own, much like we do for profiler, where we use it to capture and suppress the debug output of PSTrace. 

And then instead of triggering the prompt it would respond by some predefined value. Or possible via a mock of a function we would publish. Beacuse then we can use all the default behavior of mock (filtering, verify etc).

Something like: 

Mock Set-HostPrompt { 1 } 
& scriptThatPrompts.ps1

https://github.com/nohwnd/Profiler/blob/main/csharp/Profiler/TracerHostUI.cs#L30 

It might even work in profiling, because we would just forward through 2 tracer host ui decorators.

**Reply by nohwnd** [reply]
Sounds like my future talk on psconf :D

**Reply by fflaten** [reply]
Welcome back! 🙂 

Replacing the host was the disruptive way I tried to steer away from 😅

Felt a little risky to be part of default runtime as it's easy to corrupt the session with a cancelled run. With Profiler I expect a clean/separate pwsh-process while Pester is more likely to be used in a shared session. 

Maybe as a opt-in feature enabled using something like `Set/Update-PesterHost` and reverted with `Restore-PesterHost`.

> Sounds like my future talk on psconf :D

💯 Perfect as a blog and/or talk for advanced usage.

**Reply by nohwnd** [reply]
> Felt a little risky to be part of default runtime as it's easy to corrupt the session with a cancelled run. 

We might be able to check for variable being in scope, and otherwise just forward everything to the default host. 

> Welcome back! 🙂

haha, thanks.

**Comment by nkrishnakishor** [comment]
Thanks  @fflaten for the reply.

Suggested code works fine if the code is within the `<file>.Tests.ps1` but when I try to invoke the script from outside the `<file>.Tests.ps1` file it throws error. Below is the script file `small-script.ps1`

```
[CmdletBinding()]
param (
    [Parameter()]
    [string]
    $String
)

if ($null -eq $PSBoundParameters["String"]){
    Write-Error "Input parameter is required"
} else {
    Write-Output "Input parameter: $String"
}
```
Below is the `small-script.Tests.ps1`
```
Describe "Checking the script functionality" {
    BeforeEach {
        $ScriptFile = "C:\Users\test\Desktop\small-script.ps1"
    }
    Context "Checking parameters" {
        It "Should return error when parameter is not used" {
            { . $ScriptFile } | Should -Throw
        }
        It "Should not return error when parameter is used" {
            { . $ScriptFile -String "localhost" } | Should -Not -Throw
        }
    }
}
```
Get the error as 
```
Pester v5.4.1

Starting discovery in 1 files.
Discovery found 2 tests in 335ms.
Running tests.

Running tests from 'C:\Users\test\Desktop\small-script.Tests.ps1'
Describing Checking the script functionality
 Context Checking parameters
   [+] Should return error when parameter is not used 21ms (10ms|11ms)
   [-] Should not return error when parameter is used 14ms (13ms|1ms)
    Expected no exception to be thrown, but an exception "The term 'C:\Users\test\Desktop\small-script.ps1' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again." was thrown from C:\Users\test\Desktop\small-script.Tests.ps1:10 char:17
        +             { . $ScriptFile -String "localhost" } | Should -Not -Thro ...
        +                 ~~~~~~~~~~~.
    at { . $ScriptFile -String "localhost" } | Should -Not -Throw, C:\Users\test\Desktop\small-script.Tests.ps1:10
    at <ScriptBlock>, C:\Users\test\Desktop\small-script.Tests.ps1:10
Tests completed in 513ms
Tests Passed: 1, Failed: 1, Skipped: 0 NotRun: 0
```

Please help me how to overcome this error

**Reply by fflaten** [reply]
As the error says, it can't find the script. Make sure it exists at the given path.

**Reply by nkrishnakishor** [reply]
It exists in the mentioned path, if you check the output one of the test was executed without error but for second test it was throwing error. They both are referring to the same file.

**Reply by fflaten** [reply]
> if you check the output one of the test was executed without error 

The first test passed because you removed the expected error message in my example, `.. | Should -Throw '*missing mandatory parameters: Param2.'`. 

`Should -Throw` alone only checks for an error, which the "script not found" error fulfills.

**Comment by nkrishnakishor** [comment]
thanks for the reply @fflaten , I tried adding the message after `Should -Throw` and trigger testing it also resulted in expected output as you pointed out. I later realized I had a type in filename 🙏 

Thanks for helping me out
